### ISSUE TITLE
checkJs error for indirect JSDoc param tag

### ISSUE THREAD

**Issue by DanielRosenwasser** [Microsoft/TypeScript#19268]
TypeScript Version: 2.6.0-dev.20171015

```js
// @ts-check

function Foo() {
    /**
     * @param {string} wat
     */
    this.bar = function(wat) {

    }
}
```

Expected: No error
Actual: `JSDoc '@param' tag has name 'wat', but there is no parameter with that name.`

**Comment by DanielRosenwasser** [comment]
Probably due to #18777.

**Comment by OliverJAsh** [comment]
Any chance this will be fixed in TS 2.6? This is a big blocker for us upgrading to 2.6, as we use JSDoc in many places.

As a workaround, I'm using `@ts-ignore`.

**Comment by sandersn** [comment]
Yeah, it's certainly #18777 -- the type is correct, it just issues a spurious error.

**Comment by sandersn** [comment]
Fix is up at #19355

**Comment by OliverJAsh** [comment]
@sandersn Thanks for getting this fixed so quickly!

**Comment by KtorZ** [comment]
Hey!
I ran into a similar issue but from the other perspective. Using a destructuring assignment on the function argument, I got the following error:

```js
/**
 * @param {Object} args 
 * @param {string} args.prop1
 * @param {string} args.prop2
 */
function patate({ prop1, prop2 }) {

}
```

// JSDoc '@param' tag has name 'args', but there is no parameter with that name.
```

The problem is the same using an enum syntax:

```js
/**
 * @param {{prop1: string, prop2: string}} args 
 */
function patate({ prop1, prop2 }) {

}
```

Wouldn't it make sense not to raise this error in this case, but still be able to annotate the properties with the right types ?

**Comment by mhegazy** [comment]
@KtorZ this is tracked by https://github.com/Microsoft/TypeScript/issues/7429

**Comment by KtorZ** [comment]
Sweet, thanks :+1:
